### DISCUSSION TITLE
Hardhat Integration and CLI Documentation

### DISCUSSION THREAD

**Discussion by huseyindeniz** [evmts/tevm-monorepo#1013] [Q&A] [answered]
I discovered this repository and try to find out how I can replace TypeChain. I found the following package:

https://www.npmjs.com/package/@evmts/ethers

But it's not clear for me how to generate ts files from the contract. CLI page in documentation is under construction currently. Can someone show a quick example usage in hardhat. 

```ts
// import a contract with evmts
import {ERC20} from '@openzeppelin/contracts/token/ERC20/ERC20.sol'
import {providers} from 'ethers'
// create a provider or signer for the ethers contract
const provider = new providers.JsonRpcProvider('http://localhost:8545')
// create a typesafe contract
const contract = createEthersContract(myContract, {chainId: 1, runner: provider})
```

the example code above doesn't give a clue how can I save the typescript file. I assume some CLI command will do the trick.

**Comment by roninjin10** [comment]
Hey @huseyindeniz .Notice how in the code you shared ERC20 is being imported directly in typescript. unlike with tools like typechain you don't even need to compile the contract with hardhat or generate any typescript. It "just works".

See bundler docs here: https://tevm.sh/learn/solidity-imports/

**Reply by huseyindeniz** [reply]
That's exactly why I'm confused. Let me explain what I need. I don't want to mix my blockchain project with my frontend project. I just want to generate TypeScript files in my Hardhat project, copy them to the frontend (along with the code), and use them there. My frontend should have no knowledge of where the Solidity files are. This is how I'm currently using TypeChain. Is this possible?

**Reply by roninjin10** [reply, marked as answer]
Yes the docs linked have a section on how to do code gen. But tldr simply run “npx tevm gen . /path/to/contract.sol

Note however this is the same thing as importing the file from a different package but extra steps. Importing doesn’t tightly couple code. Imports are by definition static, happening at build time not runtime, and explicitly using a package boundary

**Reply by roninjin10** [reply]
Npx tevm has gotten a  new implementation btw very recently so use npx tevm not npx tevm-gen. Npx tevm supports globs

**Reply by huseyindeniz** [reply]
Great! I think when I wrote my first message, the codegen section was still under construction. I'll try this when I have time. Thank you very much for pointing it out! (I'll also check your note later. 'Importing doesn't tightly couple code' sounds good, but I don't fully understand it right now since I'm not working on blockchain projects these days)

**Reply by roninjin10** [reply]
You might be the second or third dev ever using this feature so reach out if you run into issues
